From a80fd56a5f9eb6b47017ac8b514e1c35aba9e583 Mon Sep 17 00:00:00 2001
From: Anson Huang <Anson.Huang@nxp.com>
Date: Tue, 5 Sep 2017 17:31:43 +0800
Subject: [PATCH 2509/5242] MLK-16372-2 cpufreq: imx8: add cpu-freq cooling
 support

commit  fb2bb3561f766286d5d208477ab00385d4229761 from
https://source.codeaurora.org/external/imx/linux-imx.git

Register cpu-freq cooling device if device tree
supports cooling-cells, different cluster can
have its own cooling device settings.

Signed-off-by: Anson Huang <Anson.Huang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/cpufreq/imx8-cpufreq.c |   22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/drivers/cpufreq/imx8-cpufreq.c b/drivers/cpufreq/imx8-cpufreq.c
index 5a45b7d..10e77d0 100644
--- a/drivers/cpufreq/imx8-cpufreq.c
+++ b/drivers/cpufreq/imx8-cpufreq.c
@@ -10,6 +10,7 @@
 #include <linux/clk.h>
 #include <linux/cpu.h>
 #include <linux/cpufreq.h>
+#include <linux/cpu_cooling.h>
 #include <linux/err.h>
 #include <linux/module.h>
 #include <linux/of.h>
@@ -30,6 +31,7 @@ struct imx8_cpufreq {
 static struct cpufreq_frequency_table *freq_table[MAX_CLUSTER_NUM];
 static unsigned int transition_latency[MAX_CLUSTER_NUM];
 struct device *cpu_dev;
+static struct thermal_cooling_device *cdev[2];
 
 static int imx8_set_target(struct cpufreq_policy *policy, unsigned int index)
 {
@@ -90,6 +92,25 @@ static int imx8_cpufreq_init(struct cpufreq_policy *policy)
 	return ret;
 }
 
+static void imx8_cpufreq_ready(struct cpufreq_policy *policy)
+{
+	struct device_node *np = of_get_cpu_node(policy->cpu, NULL);
+	unsigned int cluster_id = topology_physical_package_id(policy->cpu);
+
+	if (of_find_property(np, "#cooling-cells", NULL)) {
+		cdev[cluster_id] = of_cpufreq_cooling_register(np, policy);
+
+		if (IS_ERR(cdev[cluster_id]) && PTR_ERR(cdev[cluster_id]) != -ENOSYS) {
+			pr_err("cpu%d is not running as cooling device: %ld\n",
+					policy->cpu, PTR_ERR(cdev[cluster_id]));
+
+			cdev[cluster_id] = NULL;
+		}
+	}
+
+	of_node_put(np);
+}
+
 static struct cpufreq_driver imx8_cpufreq_driver = {
 	.flags = CPUFREQ_NEED_INITIAL_FREQ_CHECK,
 	.verify = cpufreq_generic_frequency_table_verify,
@@ -98,6 +119,7 @@ static int imx8_cpufreq_init(struct cpufreq_policy *policy)
 	.init = imx8_cpufreq_init,
 	.name = "imx8-cpufreq",
 	.attr = cpufreq_generic_attr,
+	.ready = imx8_cpufreq_ready,
 #ifdef CONFIG_PM
 	.suspend = cpufreq_generic_suspend,
 #endif
-- 
1.7.9.5

